From 1a4a1e90b747d65bdcf9e0de622776a15ed4cad4 Mon Sep 17 00:00:00 2001
From: Tom Gundersen <teg@jklm.no>
Date: Thu, 12 Jul 2018 22:14:03 +0200
Subject: [PATCH] audit: fix check for existing capability

We must not treat the return code for capng_has_capability() as a boolean,
it returns 0 if the capability is not set, 1 if it is, but CAPNG_FAIL on
failure.

Internally, it calls capng_get_caps_process() if needed, and if this fails,
the failure is forwarded.

Signed-off-by: Tom Gundersen <teg@jklm.no>
---
 src/util/audit.c | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/src/util/audit.c b/src/util/audit.c
index 5681b1c..1f73d49 100644
--- a/src/util/audit.c
+++ b/src/util/audit.c
@@ -55,14 +55,25 @@ int util_audit_drop_permissions(uint32_t uid, uint32_t gid) {
                 if (r < 0)
                         return error_origin(-errno);
         } else {
-                int have_audit_write;
+                bool have_audit_write;
+
+                r = capng_have_capability(CAPNG_PERMITTED, CAP_AUDIT_WRITE);
+                if (r == CAPNG_FAIL)
+                        return error_origin(-EIO);
+                else if (r == 1)
+                        have_audit_write = true;
+                else
+                        have_audit_write = false;
 
-                have_audit_write = capng_have_capability(CAPNG_PERMITTED, CAP_AUDIT_WRITE);
                 capng_clear(CAPNG_SELECT_BOTH);
-                if (have_audit_write)
-                        capng_update(CAPNG_ADD,
-                                     CAPNG_EFFECTIVE | CAPNG_PERMITTED,
-                                     CAP_AUDIT_WRITE);
+
+                if (have_audit_write) {
+                        r = capng_update(CAPNG_ADD,
+                                         CAPNG_EFFECTIVE | CAPNG_PERMITTED,
+                                         CAP_AUDIT_WRITE);
+                        if (r < 0)
+                                return error_origin(-EINVAL);
+                }
 
                 r = capng_change_id(uid, gid, CAPNG_DROP_SUPP_GRP);
                 if (r)
